### PR TITLE
 Algo change for evacuation

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#2589

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
When user tag to be removed host, Helix will proactively move replicas to other hosts in N -> N+1 manner.

(Write a concise description including what, why, how)

### Tests

- [X] The following tests are written for this issue:

TestInstanceOperation.java

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java [hunk 1]
```diff
@@ -128,6 +129,14 @@ private static Set<String> getActiveNodes(Set<String> allNodes, Set<String> live
     return activeNodes;
   }
 
+  private static Set<String> filterEvacuatingInstances(Map<String, InstanceConfig> instanceConfigMap,
```
Comment (zpinto) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java:132]: Would recommend renaming to `filterOutEvacuatingInstances`, as `filterEvacuatingInstances` sounds like it should be returning the EVACUATE nodes

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java [hunk 2]
```diff
@@ -128,6 +129,14 @@ private static Set<String> getActiveNodes(Set<String> allNodes, Set<String> live
...
+      Set<String> liveEnabledNodes) {
```
Comment (zpinto) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java:133]: Would recommend naming variable something more generic than `liveEnabledNodes`(maybe just `nodes`) as we could use this method somewhere else in the class later to filter out evacuating nodes. We can't guarantee nor should we guarantee that this method is always used with live and enabled nodes.

File: helix-core/src/test/java/org/apache/helix/integration/rebalancer/TestInstanceOperation.java [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+  @Test
+  public void testEvacuate() throws Exception {
+    // EV should contain all participants, check resources one by one
+    Map<String, ExternalView> assignment = getEV();
+    for (String resource : _allDBs) {
+      Assert.assertTrue(getParticipantsInEv(assignment.get(resource)).containsAll(_participantNames));
+    }
+
+    // cancel an evacuated instance
```
Comment (zpinto) [inline helix-core/src/test/java/org/apache/helix/integration/rebalancer/TestInstanceOperation.java:103]: Why does this comment say "cancel"?

File: helix-core/src/test/java/org/apache/helix/integration/rebalancer/TestInstanceOperation.java [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+      Set<String> newPAssignedParticipants = getParticipantsInEv(assignment.get(resource));
+      Assert.assertFalse(newPAssignedParticipants.contains(instanceToEvacuate));
+      Assert.assertTrue(newPAssignedParticipants.containsAll(currentActiveInstances));
+    }
+  }
+
+  @Test(dependsOnMethods = "testEvacuate")
+  public void testRevertEvacuation() throws Exception {
+
+    // evacuate an instance
```
Comment (zpinto) [inline helix-core/src/test/java/org/apache/helix/integration/rebalancer/TestInstanceOperation.java:125]: Should this one be "revert an evacuation"?

**Review by zpinto — COMMENTED** [review]
Overall logic looks good to me. Just some minor comments on naming and docstrings.

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java [hunk 5]
```diff
@@ -89,7 +90,7 @@ public static Set<String> getActiveNodes(Set<String> allNodes, Set<String> liveE
       Map<String, Long> instanceOfflineTimeMap, Set<String> liveNodes,
       Map<String, InstanceConfig> instanceConfigMap, ClusterConfig clusterConfig) {
     if (!isDelayRebalanceEnabled(clusterConfig)) {
-      return new HashSet<>(liveEnabledNodes);
+      return filterEvacuatingInstances(instanceConfigMap, liveEnabledNodes);
```
[the remark this answers is not in the document]
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java]: TFTR! Updated.

**Comment by xyuanlu** [comment]
This change is ready to be merged. Approved by @zpinto 
// Will fix the failing test in CI in next PR. Irrelevant to this PR.
Commit message: Add support for node evacuation.

**Pull request merged**